File: haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where

import Data.List ( sort )
import Data.Char (toLower)

anagramsFor :: String -> [String] -> [String]
anagramsFor word xxs = foldr ((++) . anagramFor word) [] xxs

anagramFor :: String -> String -> [String]
anagramFor word cand =
    [cand | sort lowerWord == sort lowerCand && lowerWord /= lowerCand ]
    where lowerWord = map toLower word
          lowerCand = map toLower cand

File: go/anagram/anagram.go
package anagram

import (
	"sort"
	"strings"
)

type byteSlice []byte

func Detect(subject string, candidates []string) []string {
	key := byteSlice(strings.ToLower(subject))
	sort.Slice(key, func(i, j int) bool {
		return key[i] < key[j]
	})

	var anagrams []string
	for _, candidate := range candidates {
		cand := byteSlice(strings.ToLower(candidate))
		sort.Slice(cand, func(i, j int) bool {
			return cand[i] < cand[j]
		})
		if string(key) == string(cand) && strings.ToLower(candidate) != strings.ToLower(subject) {
			anagrams = append(anagrams, candidate)
		}
	}
	return anagrams
}
